libyara/ahocorasick.c

```diff
@@ -209,19 +363,524 @@ static int _yr_ac_state_destroy(YR_AC_STATE* state)
+          new_state->matches_ref = input_state->matches_ref;
+        else
+          _yr_ac_copy_matches(new_state, input_state, arena);
+      }
+    }
+    else
+    {
+      new_state = _yr_ac_next_state_typed(
+          current_state, path->input, path->type, path->bitmap);
+      if (new_state != NULL)
+      {
+        if (YR_ARENA_IS_NULL_REF(new_state->matches_ref))
+          new_state->matches_ref = path->matches_ref;
+        else
+          _yr_ac_copy_matches(new_state, path, arena);
+      }
+    }
+
+    if (new_state == NULL)
+    {
+      if (input_state != NULL)
+        new_state = _yr_ac_create_copied_state(
+            current_state, input_state, arena);
+      else
+        new_state = _yr_ac_create_copied_state(current_state, path, arena);
+    }
+
+    state = path->first_child;
+
+    while (state != NULL)
+    {
+      _yr_ac_copy_path(state, new_state, NULL, arena);
+      state = state->siblings;
+    }
+  }
+}
+
+uint8_t num_to_bits[16] = {0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4};
+uint8_t num_to_pos[16] = {0, 1, 2, 0, 3, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0};
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// If only one bit is set in the element of array the encoded literal is return.
+// If the element is empty, 0 is returned.
+// Otherwise, the -1 is returned.
+//
+static int _yr_ac_return_literal(uint64_t number, uint8_t index)
+{
+  uint8_t bits;
+  uint8_t nibble = 0;
+  uint64_t num;
+  int result = 0;
+  int counter = 0;
+
+  num = number;
+
+  if (0 == num)
+    return result;
+
+  for (int i = 0; i < sizeof(uint64_t) * 2; i++)
+  {
+    // find last nibble
+    nibble = num & 0xf;
+    bits = num_to_bits[nibble];
+    if (bits > 1)
+      return -1;
+    else if (bits == 1)
+    {
+      // character decoding (+1 to detect bit on zero position)
+      result = (index * YR_BITMASK_SLOT_BITS) + (4 * i + num_to_pos[nibble]);
+      counter++;
+    }
+    num = num >> 4;
+  }
+
+  if (counter > 1)
+    return -1;
+  else
+    return result;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// If the class becomes literal, it returns a literal found.
+// Otherwise, it returns 0.
+//
+static int _yr_ac_class_is_literal(YR_BITMASK* bitmap)
+{
+  int bits = 0;
+  int counter = 0;
+  int literal = 0;
+  int non_literal = -1;
+
+  for (int i = 0; i < YR_BITMAP_SIZE; i++)
+  {
+    bits = _yr_ac_return_literal(bitmap[i], i);
+
+    if (bits == -1)
+      return non_literal;
+    else if (bits != 0)
+    {
+      literal = bits;
+      counter++;
+    }
+  }
+
+  if (counter == 1)
+  {
+    // _yr_ac_return_literal return literal + 1
+    return literal - 1;
+  }
+  assert(counter != 0);
+
+  return non_literal;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// Detect if two states are in a collision
+//
+static bool _yr_ac_find_conflict_states(
```

Comment:
The description for this function says that it detects if two states are in collision, but it receives four states, what's the purpose of the other two states? Which are the states that are checked for conflict?

Reply:
I added some comments to make the process a little bit clearer, hopefully. When I am changing states, I have to update the matches and children's states as well. This is, however, not easy, with the goal of not changing the results, as I wrote in the previous comment. For that reason, I had to split it into several steps.